### PR TITLE
SCC-CUF (CUDA Fortran) transformation/recipe

### PR DESCRIPTION
**Pull request by MichaelSt98**
**Single-Column-Coalesced CUDA Fortran (SCC-CUF) transformation** with 3 different types of transformation as automatic arrays are not supported in device code. 

Depends on the following pull requests:

* [Transformation: parametrise #10](https://github.com/ecmwf-ifs/loki/pull/10)
* [Transform utilities: single variable declaration #12](https://github.com/ecmwf-ifs/loki/pull/12)
* [Transform hoist variables: disable subroutines, minor fix regarding string comparison #13](https://github.com/ecmwf-ifs/loki/pull/13)

____

* for an **example/application** of this transformation/recipe see [Loki SCC-CUF #31](https://github.com/ecmwf-ifs/dwarf-p-cloudsc/pull/31) for CLOUDSC

### CODE REVIEW

File: transformations/transformations/scc_cuf.py
Comment (reuterbal) [inline transformations/transformations/scc_cuf.py:332]: `var_map[var] = var.clone(dimensions=var.dimensions[1:])`

File: transformations/transformations/scc_cuf.py
Comment (reuterbal) [inline transformations/transformations/scc_cuf.py:326]: `var_map[var] = var.clone(dimensions=var.dimensions + (routine.variable_map[block_dim.index],))`

File: transformations/transformations/scc_cuf.py
Comment (reuterbal) [inline transformations/transformations/scc_cuf.py:336]: `if call.name not in disable`

File: transformations/tests/test_scc_cuf.py
Comment (reuterbal) [inline transformations/tests/test_scc_cuf.py:215]: If you provide here `definitions=[kernel]` you can avoid the `enrich_calls` call

File: transformations/tests/test_scc_cuf.py
Comment (reuterbal) [inline transformations/tests/test_scc_cuf.py:229]: Is the code incomplete in the test sources or might #22 fix this?

**Review by reuterbal — CHANGES_REQUESTED** [review]
This is a fantastic achievement and very impressive work! Took me quite a while to go through all of it and can't claim I have understood every line of it, but overall looks very good.

In part because I was trying to understand what's going on, I've left quite a substantial amount of comments - they are largely a mixture of suggestions, nitpicks, or questions reflecting my lack of understanding.

I appreciate that this transformation will go through a few more iterations once it's being tried on other codes than CLOUDSC, so not everything may need action right away - provided you agree with the comment in the first place. I'd encourage to create an issue for anything you're not planning to work on, though, to ensure it's not forgotten.

Highest priority items for me would be to discuss the API/config interface questions raised in the comments. Because this has a [corresponding CLOUDSC PR](https://github.com/ecmwf-ifs/dwarf-p-cloudsc/pull/31) and changes would require symmetric modifications there, I think it would be good to discuss them now.

File: transformations/transformations/scc_cuf.py
Comment (MichaelSt98) [reply transformations/transformations/scc_cuf.py:200]: I agree. 
However, I thought we may need to discuss this more generally (what elemental (and/or pure) subroutines/functions are possible for which Fortran standard...) ... Maybe this could be even a utility function?

File: transformations/transformations/scc_cuf.py
Comment (MichaelSt98) [reply transformations/transformations/scc_cuf.py:434]: True, but `list(set(array))` does **not** guarantee the correct/original order. 

e.g. 

```py
array = ['a', 'b', 1, ' 2' 'd']
print(f"list(set(array)):           {list(set(array))}")
print(f"list(dict.fromkeys(array)): {list(dict.fromkeys(array))}")
```

gives 

```
list(set(array)):           ['b', 1, 'a', ' 2d']
list(dict.fromkeys(array)): ['a', 'b', 1, ' 2d']
```

File: transformations/transformations/scc_cuf.py
Comment (MichaelSt98) [reply transformations/transformations/scc_cuf.py:590]: Not my work ... see [Transformation: Single Column Coalesced](https://github.com/ecmwf-ifs/loki/blob/main/transformations/transformations/single_column_coalesced.py) ... 
Therefore, yes this should/could be an internal utility.

File: transformations/transformations/scc_cuf.py
Comment (MichaelSt98) [reply transformations/transformations/scc_cuf.py:622]: Same here, see [Transformation: Single Column Coalesced](https://github.com/ecmwf-ifs/loki/blob/main/transformations/transformations/single_column_coalesced.py)

File: transformations/transformations/scc_cuf.py
Comment (MichaelSt98) [reply transformations/transformations/scc_cuf.py:865]: No, the `!@cuf` sentinel (or rather what's after) becomes a statement whenever compiled with *nvfortran* (and with the correct flag and/or correct file extension) 

see [conditional compilation](https://docs.nvidia.com/hpc-sdk/compilers/cuda-fortran-prog-guide/#cfpg-conditional-compilation)

This was a test for me whether the other statements guarded by the sentinel work. Therefore, it is not really necessary anymore.

File: loki/frontend/ofp.py
Comment (MichaelSt98) [reply loki/frontend/ofp.py:1139]: Seems like it only works for `subroutine`s and not for `function`s ...

Comment (MichaelSt98) [reply loki/frontend/ofp.py:1139]: Parsing looks more or less like

```py
ast = ... = open_fortran_parser.parse(filepath, raise_on_error=True)
# write to file
import xml.etree.ElementTree as ET
tree = ET.ElementTree(ast)
tree.write("test.xml")
```

Parsing

```fortran
pure elemental real function f_elem(a)
    real, intent(in) :: a
    f_elem = a
end function f_elem
```

gives


<details>
<summary>test.xml</summary> 

```
<function col_begin="20" col_end="208" line_begin="1" line_end="4" name="f_elem">
      <header col_begin="36" col_end="37" line_begin="1" line_end="1">
        <function-stmt__begin addendum="begin" rule="1224" />
        <names col_begin="36" col_end="37" count="1" line_begin="1" line_end="1">
          <generic-name-list__begin addendum="list-begin" rule="102" />
          <name col_begin="36" col_end="37" id="a" line_begin="1" line_end="1">
            <generic-name-list-part addendum="part" col_begin="36" col_end="37" id="a" line_begin="1" line_end="1" rule="102" />
          </name>
          <generic-name-list addendum="list" count="1" rule="102" />
        </names>
      </header>
      <function-stmt col_begin="20" col_end="39" eos="&#10;" hasGenericNameList="true" hasSuffix="false" keyword="function" line_begin="1" line_end="1" name="f_elem" rule="1224" />
      <body col_begin="4" col_end="14" line_begin="2" line_end="3">
        <specification col_begin="4" col_end="26" declarations="1" implicits="0" imports="0" line_begin="2" line_end="2" uses="0">
          <declaration col_begin="4" col_end="26" line_begin="2" line_end="2" type="variable">
            <type col_begin="4" col_end="8" hasKind="false" hasLength="false" line_begin="2" line_end="2" name="real" type="intrinsic">
              <intrinsic-type-spec col_begin="4" col_end="8" hasKindSelector="false" keyword1="real" keyword2="" line_begin="2" line_end="2" rule="403" type="401" />
              <declaration-type-spec rule="502" type="500" udtKeyword="" />
            </type>
            <intent col_begin="17" col_end="19" line_begin="2" line_end="2" type="in">
              <intent-spec col_begin="17" col_end="19" intent="600" intentKeyword1="in" intentKeyword2="" line_begin="2" line_end="2" rule="517" />
            </intent>
            <attr-spec attr="811" attrKeyword="intent" col_begin="10" col_end="16" line_begin="2" line_end="2" rule="503" />
            <variables col_begin="24" col_end="25" count="1" line_begin="2" line_end="2">
              <entity-decl-list__begin addendum="list-begin" rule="503008" />
              <variable col_begin="24" col_end="25" hasInitialValue="false" line_begin="2" line_end="2" name="a">
                <entity-decl col_begin="24" col_end="25" hasArraySpec="false" hasCharLength="false" hasCoarraySpec="false" hasInitialization="false" id="a" line_begin="2" line_end="2" rule="503008" />
              </variable>
              <variable />
              <entity-decl-list addendum="list" count="1" rule="503008" />
            </variables>
            <type-declaration-stmt col_begin="25" col_end="26" eos="&#10;" line_begin="2" line_end="2" numAttributes="1" rule="501" />
          </declaration>
          <declaration-construct rule="207" />
          <declaration />
          <specification-part numDeclConstructs="1" numImplicitStmts="0" numImportStmts="0" numUseStmts="0" rule="204" />
        </specification>
        <statement col_begin="4" col_end="14" line_begin="3" line_end="3">
          <assignment col_begin="4" col_end="14" line_begin="3" line_end="3">
            <target col_begin="4" col_end="10" line_begin="3" line_end="3">
              <name col_begin="4" col_end="10" hasSubscripts="false" id="f_elem" line_begin="3" line_end="3" type="variable">
                <part-ref col_begin="4" col_end="10" hasImageSelector="false" hasSectionSubscriptList="false" id="f_elem" line_begin="3" line_end="3" rule="613" />
                <data-ref numPartRef="1" rule="612" />
                <designator hasSubstringRange="false" rule="603" />
                <variable rule="601" />
              </name>
            </target>
            <value col_begin="13" col_end="14" line_begin="3" line_end="3">
              <name col_begin="13" col_end="14" hasSubscripts="false" id="a" line_begin="3" line_end="3" type="ambiguous">
                <part-ref col_begin="13" col_end="14" hasImageSelector="false" hasSectionSubscriptList="false" id="a" line_begin="3" line_end="3" rule="613" />
                <data-ref numPartRef="1" rule="612" />
                <designator-or-func-ref rule="-3" />
                <primary rule="701" />
              </name>
            </value>
            <assignment-stmt rule="734" />
          </assignment>
          <action-stmt rule="214" />
        </statement>
        <statement>
          <executable-construct rule="213" />
          <execution-part rule="208" />
        </statement>
      </body>
      <end-function-stmt col_begin="0" col_end="208" eos="/var/folders/9d/hrxb93c96pq5wgdzt7cbvf9c0000gn/T/loki/ofp-651bb5527e8979b94fbaf1b0e2faeda4.f90:/var/folders/9d/hrxb93c96pq5wgdzt7cbvf9c0000gn/T/loki/ofp-651bb5527e8979b94fbaf1b0e2faeda4.f90" keyword1="end" keyword2="function" line_begin="4" line_end="4" name="f_elem" rule="1230" />
      <function-subprogram hasExePart="true" hasIntSubProg="false" rule="1223" />
      <ext-function-subprogram hasPrefix="true" rule="1101" />
    </function>
```

</details>

Thus, information/prefixes seem to be omitted...

File: transformations/tests/test_scc_cuf.py
Comment (MichaelSt98) [reply transformations/tests/test_scc_cuf.py:229]: after some debugging, the cause turned out to be something completely different... PR #27 solves this issue (and probably some other problems as well)

File: loki/bulk/scheduler.py
Comment (MichaelSt98) [reply loki/bulk/scheduler.py:46]: I agree: [issue #28](https://github.com/ecmwf-ifs/loki/issues/28#issue-1562734638)

File: transformations/transformations/scc_cuf.py
Comment (MichaelSt98) [reply transformations/transformations/scc_cuf.py:90]: True, but many things rely on single variable declarations (or rather the correct source code with the relevant variables to be declared in a single/unique variable declaration). So I can do that, but then you would have to specify that for many transformations/functions? 
Shouldn't we introduce something like troubleshooting (section/page) mentioning and explaining this thing?

File: transformations/transformations/scc_cuf.py
Comment (MichaelSt98) [reply transformations/transformations/scc_cuf.py:315]: no not giving the same results ... would need to investigate that

File: transformations/transformations/scc_cuf.py
Comment (MichaelSt98) [reply transformations/transformations/scc_cuf.py:25]: Ok, do you want me to change it for `HoistTemporaryArraysTransformationAllocatable` in `loki/transform/transform_hoist_variables.py` as well?

File: scripts/loki_transform.py
Comment (reuterbal) [reply scripts/loki_transform.py]: Yes, sounds good. Many thanks!

File: transformations/transformations/scc_cuf.py
Comment (reuterbal) [reply transformations/transformations/scc_cuf.py:695]: Thanks for the string-based selection! This docstring might need a little update now.

**Review by reuterbal — COMMENTED** [review]
Thanks for changes and comments!

Not sure what the plan is now, for this to go in first or to work on the related issues and rebase on top of them?

Leaving a few comments based on your replies.

File: loki/frontend/ofp.py
Comment (reuterbal) [reply loki/frontend/ofp.py:1153]: Yeah, I think I remember this... if you're curious, you can check what it looks like with the function wrapped in a module. If memory serves (unless I mix this up with something else), the prefix was captured but outside/before the function node. No need to fix this, though, we don't really want to commit resources to OFP support.

**Review by reuterbal — APPROVED** [review]
OK, I've addressed a few minor things directly and think this is in great shape now! Once CI is happy, GTG from my side.

File: loki/bulk/scheduler.py
Comment (mlange05) [reply loki/bulk/scheduler.py:46]: I think we can go ahead with this, and mop up the CLOUDSC config specific to SCC-CUF once we start actioning #28. Rationale is that, while not ideal in it's current state, it is functional and isolated enough to not affect other things. We should probably note this in the CLOUDSC repo tho (note to self!).

File: transformations/transformations/scc_cuf.py
Comment (mlange05) [reply transformations/transformations/scc_cuf.py:235]: Indeed, that should work better than explicit string conversion.

File: transformations/transformations/scc_cuf.py
Comment (mlange05) [reply transformations/transformations/scc_cuf.py:489]: I'm confused by this: `size_expressions` are declaration dimensions, while `loop.variable` should be a `dim.index`. We don't have indexing for dimensions yet, so this might be a necessary workaround for now.

**Review by mlange05 — APPROVED** [review]
Ok, first of all, this is an incredibly complicated transformation, and very impressively and very well implemented! There are still a few small comments, but these are largely cosmetic, and I can take care of them before merging if you like (this has been hanging long enough). Very cool stuff, and very impressive work!

In addition, there is a bunch of regular patterns, and commonalities with the other SCC transformations, that will require a bunch of refactoring to streamline this all a little. However, none of this should diminish what a fantastic achievement this is! So, I'm happy for this to go in, and resolve all remaining idiosyncrasies in follow-up work.

**Comment by mlange05** [comment]
Ok, I pushed a last round of code minor cleanup changes, including stuff that should avoid the repeated manual string comparison when traversing call lists. I've also confirmed that this works with the corresponding CLOUDSC PR, and will merge once CI comes back clean.

**Pull request merged**
